tests/mocks/core/services/request/RequestMock.php

```diff
@@ -0,0 +1,211 @@
+<?php
+
+namespace EventEspresso\tests\mocks\core\services\request;
+
+use EventEspresso\core\domain\services\contexts\RequestTypeContextCheckerInterface;
+use EventEspresso\core\services\request\Request;
+
+/**
+ * Class RequestMock
+ * Description
+ *
+ * @package EventEspresso\tests\mocks\core\services\request
+ * @author  Brent Christensen
+ * @since   $VID:$
+ */
+class RequestMock extends Request
+{
+
+    /**
+     * $_GET parameters
+     *
+     * @var array $get
+     */
+    public $get;
+
+    /**
+     * $_POST parameters
+     *
+     * @var array $post
+     */
+    public $post;
+
+    /**
+     * $_COOKIE parameters
+     *
+     * @var array $cookie
+     */
+    public $cookie;
+
+    /**
+     * $_SERVER parameters
+     *
+     * @var array $server
+     */
+    public $server;
+
+    /**
+     * $_REQUEST parameters
+     *
+     * @var array $request
+     */
+    public $request;
+
+    /**
+     * @var RequestTypeContextCheckerInterface
+     */
+    public $request_type;
+
+    /**
+     * IP address for request
+     *
+     * @var string $ip_address
+     */
+    public $ip_address;
+
+    /**
+     * @var string $user_agent
+     */
+    public $user_agent;
+
+    /**
+     * true if current user appears to be some kind of bot
+     *
+     * @var bool $is_bot
+     */
+    public $is_bot;
+
+
+    public function __construct(array $get, array $post, array $cookie, array $server, $ip_address = '0.0.0.0')
+    {
+        $this->get = $get;
+        $this->post = $post;
+        $this->cookie = $cookie;
+        $this->server = $server;
+        $this->request = array_merge($this->get, $this->post);
+        $this->ip_address = $this->visitorIp($ip_address);
+        parent::__construct($get, $post, $cookie, $server);
+    }
+
+
+    /**
+     * @param string $key
+     * @param string $value
+     * @param bool   $override_ee
+     * @return    void
+     */
+    public function setRequestParam($key, $value, $override_ee = false)
+    {
+        $this->request[ $key ] = $value;
+    }
```

Comment:
This basically means any calls to this method in tests using the mock are completely ignoring the conditions present in the actual code.  IMO it makes the test fragile.  I think it'd better to not override this method.

Reply:
The entire point of the mock is to remove as much of the mocked class from tests as possible, that's why this is overriding the complex logic that is in the real `Request` class with something more straightforward. It does mean that you have to be aware of what you are doing and ensuring that you are **only** testing your targeted code as opposed to performing integration tests which is what a lot of our "unit" tests have unfortunately been.

You may have noticed that the `RequestTypeContextDetectorTest` test case in this branch does not extend `EE_UnitTestCase` nor `WP_UnitTestCase`, but extends `PHPUnit_Framework_TestCase` directly because it is intended to be a pure unit test and not an integration test.

Bottom line is that the intent here is solely to unit test the `RequestTypeContextDetector` class and not the `Request` class. 
IF your intent is to perform integration tests, then you should not use mocks of any classes involved.